WIP: Allow roles to be specified for Keycloak

khawaga opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

<!--- Describe your changes in detail -->
This PR adds the ability to specify specific roles for use with Keycloak, which offers more granular control compared to groups.

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->
The motivation behind this feature was that a similar project, [Louketo Proxy](https://github.com/louketo/louketo-proxy), [has reached EOL](https://www.keycloak.org/2020/08/sunsetting-louketo-project.adoc.html), and that project had the ability to specify roles which was important for several applications. OAuth2 Proxy would be a suitable replacement if it were to offer that same feature.
<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
The environment consisted of a Kubernetes cluster with a live Keycloak server.
I have tested the following:
- Not providing the new `keycloak-roles` flag to ensure that previous functionality is maintained. 
- Providing a role that my test user did not have. User was not granted access.
- Providing a realm role that my test user had. User was granted access.
- Providing a list consisting of two roles, one of which my test user did not have. User was not granted access. This is the desired behaviour as all roles need to exist in order for a user to be granted access.
- Providing a client role that my test user had. User was granted access.
<!--- see how your change affects other areas of the code, etc. -->


## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

NickMeves commented on contrib/oauth2-proxy_autocomplete.sh:
```diff
@@ -24,7 +24,7 @@ _oauth2_proxy() {
 			return 0
 			;;
-		--@(http-address|https-address|redirect-url|upstream|basic-auth-password|skip-auth-regex|flush-interval|extra-jwt-issuers|email-domain|whitelist-domain|trusted-ip|keycloak-group|azure-tenant|bitbucket-team|bitbucket-repository|github-org|github-team|github-repo|github-token|gitlab-group|github-user|google-group|google-admin-email|google-service-account-json|client-id|client_secret|banner|footer ...
+		--@(http-address|https-address|redirect-url|upstream|basic-auth-password|skip-auth-regex|flush-interval|extra-jwt-issuers|email-domain|whitelist-domain|trusted-ip|keycloak-group|keycloak-roles|azure-tenant|bitbucket-team|bitbucket-repository|github-org|github-team|github-repo|github-token|gitlab-group|github-user|google-group|google-admin-email|google-service-account-json|client-id|client_secre ...
```
Ha - I didn't even realize this was a thing.  I bet this way out of date, I've seen many PRs adding flags that didn't touch this 🤣

khawaga replied:
I did not know that 😂, I just did a search for one of the existing flags and that's how I found this! :)

NickMeves commented on providers/keycloak.go:
```diff
@@ -95,5 +128,29 @@ func (p *KeycloakProvider) GetEmailAddress(ctx context.Context, s *sessions.Sess
 		}
 	}
 
+	if len(p.Roles) > 0 {
```
Can you refactor this into another method that does the role authorization? This addition makes this `GetEmailAddress` method too complex: https://codeclimate.com/github/oauth2-proxy/oauth2-proxy/pull/767

khawaga replied:
Sure thing, but I believe after refactoring block into its own method it would still exceed the number of allowed returns (it would become 1 less, leaving it at 5, which still exceeds the 4 allowed) so that would not fix the issue. Would you still suggest to refactor it?

JoelSpeed replied:
Code climate is advisory, take it with a pinch of salt. Would be good still to have a dedicated role check function so basically just extract the contents of this if to a function I would say

khawaga replied:
Got it, I've extracted the logic into a separate function.

NickMeves commented on providers/keycloak.go:
```diff
@@ -95,5 +128,29 @@ func (p *KeycloakProvider) GetEmailAddress(ctx context.Context, s *sessions.Sess
+
+		var roles = ExtractRolesFromClaims(claims)
+
+		if !isSubarray(roles, p.Roles) {
```
So we might need to think about how this aligns with other implementations. In general, in other group/role AuthZ logic in other providers, we've made this a if any roles match the list, access is allowed (instead of all must be present).

khawaga replied:
Oh, are roles being used in any other provider? The only thing I've seen is groups, and for that it makes sense that access would be granted for a user belonging to any specified group. This was based on the Louketo Proxy implementation.

> By default, the roles defined on a resource perform a logical AND so all roles specified must be present in the claims, this behavior can be altered by the require-any-role option, however, so as long as one role is present the permission is granted.

I am open for suggestions though, would you like me to change it so that it grants access if any of the provided roles are present?

NickMeves commented on providers/keycloak.go:
```diff
@@ -63,12 +66,42 @@ func (p *KeycloakProvider) SetGroup(group string) {
+	p.Roles = roles
+}
+
+func ExtractRolesFromClaims(claims map[string]interface{}) []string {
```
Can you add Unit Tests for this new method? (And add new cases for the additions you made to existing methods)

khawaga replied:
Sure thing, would it make sense to change [auth_test.go](https://github.com/oauth2-proxy/oauth2-proxy/blob/master/providers/auth_test.go#L13) in order to allow passing an access token when calling `CreateAuthorizedSession`? Since the roles are extracted from the access token so it needs to be valid otherwise token parsing will fail.

khawaga replied:
I've added two more tests.

khawaga replied:
I've edited the token in auth_test.go in order to contain a valid token instead of the dummy that was being used. It's currently hardcoded which is not exactly ideal, but if you'd like I can add another method to create the token (that can then be called from the tests in order to create and then set it).

JoelSpeed replied:
Does this function need to be exported?

khawaga replied:
It does not, I unexported it :)

NickMeves reviewed (changes requested):
Thanks for submitting this!  I left some notes to try to get this aligned with the other group/role work that is ongoing in other providers.

khawaga commented:
@NickMeves I've added some tests and I've also replied to your comments. Thank you for the prompt review!

JoelSpeed commented on providers/auth_test.go:
```diff
@@ -8,7 +8,8 @@ import (
 )
 
-var authorizedAccessToken = "imaginary_access_token"
+// The following is a valid JWT, it can be decoded, edited and re-encoded using https://jwt.io/
+var authorizedAccessToken = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkhvbWVyIFNpbXBzb24iLCJlbWFpbCI6IkNodW5reUxvdmVyNTNAYW9sLmNvbSIsImlhdCI6MTUxNjIzOTAyMiwicmVhbG1fYWNjZXNzIjp7InJvbGVzIjpbInRlc3QtcmVhbG1yb2xlMSIsInRlc3QtcmVhbG1yb2xlMiJdfSwicmVzb3VyY2VfYWNjZXNzIjp7ImNsaWVudCI6eyJyb2xlcyI6WyJ0ZXN0LWNsaWVudHJvbGUxIiwidGVzdC1jbGllbnRyb2xlMiJdfX19.DqKfcantBn7B8acd9rl0 ...
```
What key was used to sign this?

khawaga replied:
It's using the default key on jwt.io which is "your-256-bit-secret" 😅
Should I change it to something else? Or add it as a comment?

NickMeves replied:
I'm centralizing OIDC logic in this PR and making sample IDTokens more accessible to test cases across providers here: https://github.com/oauth2-proxy/oauth2-proxy/pull/936/files#diff-e719a8f54d2fe3115f6fcaabd056be5b53e86a3f04004d3c53f0b6c5b0409d68R34

Might be of use to this PR if you need valid IDTokens for testing.

JoelSpeed commented on providers/keycloak.go:
```diff
@@ -63,12 +66,42 @@ func (p *KeycloakProvider) SetGroup(group string) {
 		SetHeader("Authorization", "Bearer "+s.AccessToken).
 		Do().
 		UnmarshalJSON()
+
```
Any reason for this new line?

khawaga replied:
I removed it.

JoelSpeed commented on providers/keycloak.go:
```diff
@@ -95,5 +128,29 @@ func (p *KeycloakProvider) GetEmailAddress(ctx context.Context, s *sessions.Sess
+		claims := make(map[string]interface{})
+
+		// Decode JWT token without verifying the signature
+		token, err := jwt.ParseSigned(s.AccessToken)
+
+		if err != nil {
```
Would prefer not to have a new line between the err declaration and check if the err happened

khawaga replied:
Got it!

JoelSpeed commented on providers/keycloak.go:
```diff
@@ -95,5 +128,29 @@ func (p *KeycloakProvider) GetEmailAddress(ctx context.Context, s *sessions.Sess
+			logger.Printf("failed to parse claims %s", err)
+		}
+
+		var roles = ExtractRolesFromClaims(claims)
+
+		if !isSubarray(roles, p.Roles) {
```
Nit
```suggestion
		roles := ExtractRolesFromClaims(claims)
		if !isSubarray(roles, p.Roles) {
```

khawaga replied:
Thank you for the suggestion, I changed it :)

JoelSpeed commented on providers/util.go:
```diff
@@ -29,3 +29,20 @@ func makeOIDCHeader(accessToken string) http.Header {
+
+// Return true if secondArray is a subset of firstArray
+func isSubarray(firstArray, secondArray []string) bool {
+	arraySet := make(map[string]bool)
```
Nit, better to use `map[string]struct{}` and then `arraySet[value] = struct{}{}` when setting it and `if _, found := arraySet[value]; !found` when checking, this is more explicit and idiomatic in Go.

khawaga replied:
I truly appreciate the tip! I changed it now.

khawaga commented:
@NickMeves I've done the requested changes and I've also replied to your comments, thanks!

NickMeves reviewed (changes requested):
Sorry if these comments didn't get submitted before... I guess I just left them pending somehow and didn't get them over to you. Forgive me if some are outdated at this point -- but I see a few issues that might still need to be resolved.

khawaga commented:
@NickMeves I've changed the logic (and the docs) to use ANY instead of ALL for role-matching, and I've also added some OK checks to the type assertions. Are there any other issues that need to be resolved?

NickMeves commented:
Adding in some minor scope creep:

PR #616 which just merged did very similar AuthZ work as this for the OIDC provider.  It additionally added in a global `Groups` field to all sessions that when set would result in `X-Forwarded-Groups` HTTP header to backend applications (so they could use it for fine-grained route based RBAC themselves):

https://github.com/oauth2-proxy/oauth2-proxy/pull/616/files#diff-6fd8df33d9f8086bc31e28c375fbc0abR951

I think we should dump the roles you extract into the session so users get access to these details in their backend applications easily.

I know the names don't fully line up, but the bazillion providers all have different group-like concepts with different names -- so I think `Groups` is the ideal area to centralize them letting our first class provider OIDC get the naming that mirrors its conventions.

NickMeves commented:
> @NickMeves I've changed the logic (and the docs) to use ANY instead of ALL for role-matching, and I've also added some ok checks to the type assertions. Are there any other issues that need to be resolved?

I'll give everything another look this afternoon.  For now just take care of the code climate findings: https://codeclimate.com/github/oauth2-proxy/oauth2-proxy/pull/767

Are these the same findings we talked about before that might've been challenging to refactor? Or are these new?

NickMeves commented:
So I did some work this weekend to cleanup the session enrichment process after token redeem (https://github.com/oauth2-proxy/oauth2-proxy/pull/796) & centralizing authorization logic to be more global and uniform across providers (https://github.com/oauth2-proxy/oauth2-proxy/pull/797) since we are having so many authorization PRs lately taking their providers in different directions.

This PR caught my eye since it looks like the previous Keycloak roles logic was just tossed into `GetEmailAddress` since no better place existed even though it ideally shouldn't be there.  The authZ logic also looks very similar to other places, so hopefully it can just hook into the default `Authorize` provider method (or implement its own and have it called at the right moments by `oauthproxy`.

I'll check back once those PRs merge.

NickMeves commented:
Sorry for the delay.  If you still want to get this over the finish line:

https://github.com/oauth2-proxy/oauth2-proxy/pull/797 merged.

The refactor to `EnrichSessionState` & `Authorize` provider interface methods are now in master. This should be completely unblocked now.

khawaga commented:
@NickMeves I would still like to see this through, definitely.
I'll take a look at the required changes and refactor. Thanks for the heads-up!

bakito commented on CHANGELOG.md:
```diff
@@ -12,6 +12,7 @@
 
 - [#764](https://github.com/oauth2-proxy/oauth2-proxy/pull/764) Document bcrypt encryption for htpasswd (and hide SHA) (@lentzi90)
 - [#616](https://github.com/oauth2-proxy/oauth2-proxy/pull/616) Add support to ensure user belongs in required groups when using the OIDC provider (@stefansedich)
+- [#757](https://github.com/oauth2-proxy/oauth2-proxy/pull/767) Add ability to specify roles when using the Keycloak provider (@khawaga)
```
You have a typo here. Should be #767

khawaga replied:
Thanks for the heads-up! I fixed it!

NickMeves commented:
As a heads up, we are likely looking to make a V7 release sometime in mid-January.  Just wanted to give you a timeline update in case you were looking to get this in for the next release.

NickMeves commented on pkg/apis/options/options.go:
```diff
@@ -37,6 +37,7 @@ type Options struct {
 
 	AuthenticatedEmailsFile  string   `flag:"authenticated-emails-file" cfg:"authenticated_emails_file"`
 	KeycloakGroup            string   `flag:"keycloak-group" cfg:"keycloak_group"`
+	KeycloakRoles            []string `flag:"keycloak-roles" cfg:"keycloak_roles"`
```
```suggestion
	KeycloakRoles            []string `flag:"keycloak-role" cfg:"keycloak_roles"`
```

Flags are singular, configs are plural in multi-param cases.

NickMeves commented on pkg/apis/sessions/session_state.go:
```diff
@@ -27,6 +27,7 @@ type SessionState struct {
 	Email             string   `msgpack:"e,omitempty"`
 	User              string   `msgpack:"u,omitempty"`
 	Groups            []string `msgpack:"g,omitempty"`
+	Roles             []string `msgpack:"r,omitempty"`
```
Nothing new should be added to the session. You should add the roles to the existing `session.Groups` option to ride default `Authorize` & header injection implementations. This helps keep the codebase more uniform so providers can automatically get new features as we adopt them globally if they adhere to conventions.

When providers have multiple group-like structures that might want to be in this field, we've adopted prefixing the names.  Here's a sample from a very similar GitLab PR that just merged: https://github.com/oauth2-proxy/oauth2-proxy/pull/630/files#diff-9179d3c8d617839cfc99bbcd5a63a315b6cf3e652bb7f7fb9dd4bd8c6b40128aR310

khawaga replied:
Got it, I will make the required changes.

NickMeves commented on providers/keycloak.go:
```diff
@@ -63,37 +68,151 @@ func (p *KeycloakProvider) SetGroup(group string) {
+	return roles, nil
+}
+
+func (p *KeycloakProvider) isAuthorizedGroupMember(s *sessions.SessionState) (bool, error) {
```
This should be unneeded if you set the roles you want like this to take advantage of the global `Authorize` logic. Use a prefixed name like I mentioned above to not collide with groups -- e.g. `role:foorbar`

https://github.com/oauth2-proxy/oauth2-proxy/blob/d67d6e3152e2a1c9614906ee9bd9890f8da9a007/pkg/validation/options.go#L289

Here are the pertinent areas that do this for you:
https://github.com/oauth2-proxy/oauth2-proxy/blob/d67d6e3152e2a1c9614906ee9bd9890f8da9a007/providers/provider_data.go#L56

https://github.com/oauth2-proxy/oauth2-proxy/blob/d67d6e3152e2a1c9614906ee9bd9890f8da9a007/providers/provider_default.go#L102

khawaga replied:
I've done that, I would still appreciate your input on my comment regarding the groups/roles ANY logic on the other thread.

NickMeves commented on providers/keycloak.go:
```diff
@@ -63,37 +68,151 @@ func (p *KeycloakProvider) SetGroup(group string) {
+	return true, nil
+}
+
+func (p *KeycloakProvider) hasRequiredRole(s *sessions.SessionState) (bool, error) {
```
This is also unneeded in favor of the new `SetAllowedGroups` & `Authorize` functionality

khawaga replied:
Got it, looking at that now that actually makes this much simpler and cleaner. One part I'd like to discuss here though, I've noticed that this would effectively result in an ANY behaviour between roles and groups which may not fit here, I would appreciate your input on this. I've seen that this was already discussed here, although in a different context: https://github.com/oauth2-proxy/oauth2-proxy/pull/630/files#r529172740

NickMeves replied:
For the case where `--keycloak-groups` & `--keycloak-roles` is set, if you want to enforce that a user has at least 1 group AND at least 1 role that matches, you should override the default `Authorize` method in the Keycloak provider.

I'd imagine you'd manage 2 sets (groups and roles) and force membership checks in BOTH sets to return true to result in `Authorize` returning true.

khawaga replied:
I think that would make more sense, what do you think? Should I go ahead and do that then?

NickMeves replied:
Yeah, if you need separate ANY logic between groups and roles, and between the sets AND logic, overriding the default with a custom `Authorize` method that does that logic is best.

NickMeves commented on providers/keycloak.go:
```diff
@@ -63,37 +68,151 @@ func (p *KeycloakProvider) SetGroup(group string) {
+	return email, nil
+}
+
+func getGroups(userInfo *simplejson.Json) ([]string, error) {
```
Will we need to harmonize this groups scope creep with this related PR if it gets picked up again? https://github.com/oauth2-proxy/oauth2-proxy/pull/861

khawaga replied:
I believe so.

NickMeves commented on providers/keycloak.go:
```diff
@@ -63,37 +68,151 @@ func (p *KeycloakProvider) SetGroup(group string) {
+		return nil, fmt.Errorf("failed to parse token: %v", err)
+	}
+	// Parse claims
+	if err := token.UnsafeClaimsWithoutVerification(&claims); err != nil {
```
You've gotta `Verify` tokens if we want to use any data from them as part of the session. I'm gonna be a stickler for that for security reasons.

We've made `Verifier *oidc.IDTokenVerifier` an object available globally that any providers that set it can use.
https://github.com/oauth2-proxy/oauth2-proxy/blob/d67d6e3152e2a1c9614906ee9bd9890f8da9a007/pkg/validation/options.go#L237

Take a look at this PR and maybe wait until it merges to see if its new global OIDC helpers make sense for Keycloak and how you extract claims: https://github.com/oauth2-proxy/oauth2-proxy/pull/936

Keycloak, Gitlab & Azure all have OIDC elements, that PR's goal is to unify the logic in how we turn IDToken claims into Session fields.

NickMeves commented on providers/keycloak.go:
```diff
@@ -63,37 +68,151 @@ func (p *KeycloakProvider) SetGroup(group string) {
+func extractRolesFromClaims(claims map[string]interface{}) ([]string, error) {
+	var roleList []string
+
+	if realmRoles, found := claims["realm_access"].(map[string]interface{}); found {
```
By chance do you have a sample of what Keycloak IDToken claims are present? I'm curious if some of the data pulled via the userinfo/profileURL endpoint can potentially be extracted from IDToken claims and we don't need that extra HTTP call.

NickMeves replied:
(This is definitely scope creep and not needed for your PR -- I'm just curious for my own understanding).

khawaga replied:
Not sure about the IDToken, but the AccessToken contains both the email and the groups which should be sufficient to get the info used here. I would say that HTTP call is not particularly necessary.

khawaga replied:
Although that might require a breaking change since I believe that's configurable, if groups are returned via userinfo but not included in the token.

khawaga replied:
I checked, and it should possible to get the required info from the IDToken as well with the right configuration.

NickMeves replied:
Oh interesting... that would definitely make this different than other providers that are more OIDC based and get extra claims from an IDToken we can verify.

Does Keycloak have any documentation on where to get the public key to verify the signature on the Access Tokens?

I really only want to use claims extracted from an unverified token as a last resort. There's gotta be a way we can verify the signature of the token?

NickMeves replied:
Is this an option? https://stackoverflow.com/questions/54318633/getting-keycloaks-public-key

khawaga replied:
Keycloak has a JWKS endpoint at `/auth/realms/<realm_name>/protocol/openid-connect/certs`, it should be possible to verify it using that, should I reuse `oidc-jwks-url` for this purpose? If so, what would be the best way to pass it to the provider?

NickMeves replied:
So I need your help explaining something 😅 

Keycloak as a provider confuses me. When I look at it, it has barebones implementations as if it were vanilla OAuth2. (and should what it calls the `validateURL` be the `profileURL`...).

BUT - I get sample configs from users using OIDC config options with it set as the provider in questions frequently. If it is truly OIDC compliant in all cases, I feel like it should be baked into the provider more.

This way we can take advantage of all the new OIDC based features we are making global and Keycloak can take advantage of them (bearer token to session logic, generic claims to session fields - future adhoc claims -> session -> header support).

The reason Keycloak would still be a distinct provider instead of just using Keycloak with an OIDC provider: your custom roles from access tokens logic here, any special quirks in the schema of the userinfo endpoint etc.

Global OIDC logic refactor: https://github.com/oauth2-proxy/oauth2-proxy/pull/936

CC: @JoelSpeed -- if you have some history context or experience with this provider yourself, I would love your thoughts.

khawaga replied:
My understanding is that it is OIDC-compliant (it's OpenID-certified: https://openid.net/developers/certified/), I believe validateURL should be userInfoURL.
It _should_ work with the standard OIDC provider implementation (although I have not tried that myself), but yeah basically the additional role functionality is probably worth having it as an independent provider.

NickMeves replied:
So I think to get at an OIDC IDToken Verifier (if that works on the Access Token) or underlying raw JWKS, we use the existing flags to set it: https://oauth2-proxy.github.io/oauth2-proxy/docs/configuration/oauth_provider/#skip-oidc-discovery

During options validation, if you see `--keycloak-roles` you'd need to assert you also have the OIDC URLs setup so you have the tools at your disposal to verify the signature of the Access Token.

NickMeves replied:
And on the userinfo... I guess technically it works as a validateURL since it will take an access token and return 200, which is all this is looking for: https://github.com/oauth2-proxy/oauth2-proxy/blob/d67d6e3152e2a1c9614906ee9bd9890f8da9a007/providers/internal_util.go#L49

But it is primarily used as a profileURL in the Keycloak provider itself.  I'll see if I can get a PR up to clean that up in a backwards compatible way.

NickMeves reviewed (changes requested):
So I think all my refactors since your originally submitted this have changed the lay of the land a bit 😅 

I've tried to leave notes on how to integrate with the new standardized/global ways we've setup to do this. Please ask whatever question you need, I'll try to respond quickly to get you going in the right direction with some of the new architectures.

I haven't looked at the new test cases at all -- I'll review those once we get the code aligned to the new standards.

NickMeves commented on providers/keycloak.go:
```diff
@@ -63,37 +67,126 @@ func (p *KeycloakProvider) SetGroup(group string) {
+	if err != nil {
+		logger.Printf("failed to extract client roles: %s", err)
+	}
+	return append(customClaims.RealmAccess.Roles, clientRoles...), nil
```
It feels weird to pollute & extend the underlying object we extracted from the token claims.

I understand the performance motivations of reusing the existing slice since it is done at the end of the function anyway, but since these are small & therefore performance is negligible, maybe clarity is more ideal? Took me a second on first glance to realize the desired output was intended as `RealmAccess` + ResourceAccess`

khawaga replied:
You're right, I will change this for better clarity.

NickMeves commented on providers/keycloak.go:
```diff
@@ -63,37 +67,126 @@ func (p *KeycloakProvider) SetGroup(group string) {
+	for name, list := range resourceAccess {
+		scopes, ok := list.(map[string]interface{})
+		if !ok {
+			return nil, errors.New("error parsing client roles from claims")
```
Do you want to completely error if only 1 element in the list had parsing errors or skip it and make a log message?

OIDC complex claims code went the second route: https://github.com/oauth2-proxy/oauth2-proxy/blob/d67d6e3152e2a1c9614906ee9bd9890f8da9a007/providers/oidc.go#L294

NickMeves commented on providers/keycloak.go:
```diff
@@ -63,37 +67,126 @@ func (p *KeycloakProvider) SetGroup(group string) {
+			return nil, errors.New("error parsing client roles from claims")
+		}
+		if roles, found := scopes["roles"]; found {
+			for _, r := range roles.([]interface{}) {
```
Do you need to handle when this isn't a `[]interface{}` or have fallback code that somehow marshals non-slices if you get a singleton? We have to handle that case in the OIDC side: https://github.com/oauth2-proxy/oauth2-proxy/pull/936/commits/48baecb5c996d378f6871fa41a1c4824354dd083

khawaga replied:
From what I've seen it always returns a slice even when if it only contains a single element, is there any other case that I'm missing?

NickMeves replied:
Cool - that makes sense that it is uniform since it is all provided by the singular Keycloak. OIDC is all over the place cause identity providers do some crazy stuff with `groups` since it isn't a totally official claim.

NickMeves commented on providers/keycloak.go:
```diff
@@ -63,37 +67,126 @@ func (p *KeycloakProvider) SetGroup(group string) {
+		}
+		if roles, found := scopes["roles"]; found {
+			for _, r := range roles.([]interface{}) {
+				clientRoles = append(clientRoles, fmt.Sprintf("%s:%s", name, r))
```
Will `name` always be `role`? Should we make that prefix explicit here instead of from a variable so we don't get surprised later?

NickMeves replied:
Oh wait, nevermind - I see the `role` prefix later.  So this has a different `name:role` structure that represents the full role name?

khawaga replied:
Yes, so effectively this would end up in allowed groups as role:clientname:clientrole. These are client roles which are namespaced by client, so while the structure is predictable the client name could be anything, hence the additional prefix.

khawaga commented:
@NickMeves since #953 is now merged I'll rebase my branch on top of that, I'll post an update here once that's done.

NickMeves commented:
> @NickMeves since #953 is now merged I'll rebase my branch on top of that, I'll post an update here once that's done.

So take a look at this issue: https://github.com/oauth2-proxy/oauth2-proxy/issues/956

Long term, there's a few providers that I want to move from being OAuth based to OIDC based to embrace all that functionality. Keycloak is one of them (especially with the features you want to add). The existing `api` scope OAuth provider sorta sucks and I've realized alot of Keycloak users are hooking it into the `oidc` provider (after getting burned by thinking `keycloak` provider is for them but is missing functionality).

So after working on the existing `keycloak` provider. I think it eventually needs to get deprecated in favor of an OIDC based variant.

I think we make a new `keycloak-oidc` provider for the time being (in v8 it will replace the `keycloak`). We can have it be based on the `OIDCProvider` as a base. Then I think you get nearly all the functionality you need in the OIDC methods and you just need to add a few methods that extracts the roles from the Access Token.

That also gets you the OIDC supporting methods to appropriately verify your Access Token before extracting claims from it.

What do you think?

Let me know if you need a hand spiking out what a base `keycloak-oidc` provider would look like.

khawaga commented:
@NickMeves what you're suggesting makes perfect sense, is there currently any provider that's following this pattern (using the OIDC provider as a base, then extending it with additional functionality) that I can look at? I would appreciate your input on this. :)

NickMeves commented:
> @NickMeves what you're suggesting makes perfect sense, is there currently any provider that's following this pattern (using the OIDC provider as a base, then extending it with additional functionality) that I can look at? I would appreciate your help on this. :)

@khawaga You would be the first - the trailblazer 😉 

Let me know if you need a hand. I can find time over the weekend to spike out a baseline `keycloak-oidc` provider that would completely inherit `OIDCProvider` methods (it would have `OIDCProvider` in its struct composition instead of `ProviderData` in essence).

Then from that you could add on the few overrides for things like `EnrichSession` to handle the custom Access Token role logic.

NickMeves commented:
I will finish this work in #1107 since this PR has gone stale.
